persistence/sql/persister_identity.go

```diff
@@ -565,3 +598,103 @@ func (p *Persister) injectTraitsSchemaURL(ctx context.Context, i *identity.Ident
 	i.SchemaURL = s.SchemaURL(p.r.Config(ctx).SelfPublicURL()).String()
 	return nil
 }
+
+func (p *Persister) getJSONSearchQuery(ctx context.Context, field string, values []string) pop.ScopeFunc {
+	return func(q *pop.Query) *pop.Query {
+		// Removing this would allow SQL Injection. Never remove this.
+		if !validFields.MatchString(field) {
+			return q
+		}
+
+		prefix, key := extractFieldAndInnerFields(field)
+
+		switch p.Connection(ctx).Dialect.Name() {
+		case "sqlite3", "mysql", "mariadb":
+			for _, value := range values {
+				if key == "" {
+					q = q.Where(fmt.Sprintf(`json_extract(%s, '$') = ?`, prefix), value)
+				} else {
+					q = q.Where(fmt.Sprintf(`json_extract(%s, '$.%s') = ?`, prefix, key), value)
+				}
+			}
+			return q
+		case "postgres", "cockroach":
+			for _, value := range values {
+				param := fmt.Sprintf(`%s`, value)
+				if key != "" {
+					param = fmt.Sprintf(`{"%s":"%s"}`, key, value)
+				}
+				q = q.Where(fmt.Sprintf(`%s @> ?`, prefix), param)
+			}
+			return q
+		}
+
+		return q
+	}
+}
+
+func (p *Persister) searchCredentialQuery(field string, values []string) pop.ScopeFunc {
+	var innerField string
+	_, innerField = extractFieldAndInnerFields(field)
+	switch innerField {
+	case "type":
+		return func(q *pop.Query) *pop.Query { return q.Where("credential_types.name IN (?)", values) }
+	case "identifier":
+		return func(q *pop.Query) *pop.Query { return q.Where("credential_identifiers.identifier IN (?)", values) }
+	default:
+		return func(q *pop.Query) *pop.Query { return q }
+	}
+}
+
+func (p *Persister) buildIdentitySearchScope(ctx context.Context, filters map[string][]string) pop.ScopeFunc {
+	return func(q *pop.Query) *pop.Query {
+		for field, values := range filters {
+			// Removing this would allow SQL Injection. Never remove this.
+			if !validFields.MatchString(field) {
+				return q
+			}
+
+			if !p.validateFields(field) {
+				continue
+			}
+
+			if strings.HasPrefix(field, "traits.") {
+				q = q.Scope(p.getJSONSearchQuery(ctx, field, values))
+				continue
+			}
+
+			if strings.HasPrefix(field, "credentials") {
+				q = q.Scope(p.searchCredentialQuery(field, values))
+				continue
+			}
+
+			q = q.Where(fmt.Sprintf("%s IN (?)", field), values)
+		}
+		return q
+	}
+}
+
+func extractFieldAndInnerFields(field string) (string, string) {
+	if !strings.Contains(field, ".") {
+		return field, ""
+	}
+	dotIndex := strings.Index(field, ".")
+	return field[:dotIndex], field[dotIndex+1:]
+}
+
+var validFields = regexp.MustCompile(`[a-zA-Z0-9._]+`)
+
+func (p *Persister) validateFields(field string) bool {
+	return validFields.MatchString(field)
+}
+
+var validValues = regexp.MustCompile(`[%]+`)
+
+func (p *Persister) validateValues(values []string) bool {
+	for _, value := range values {
+		if validValues.MatchString(value) {
+			return false
+		}
+	}
+	return true
+}
```

Comment:
These are complex functions but they do not have any tests. Can you please add unit tests to ensure that these behave the way you want?